discord/responses.go

```diff
@@ -84,6 +84,34 @@ func (guild *GuildState) linkPlayerResponse(s *discordgo.Session, GuildID string
 	}
 }
 
+func (guild *GuildState) linkPlayerResponseByIdAndColor(s *discordgo.Session, userID string, color string) {
```

Comment:
The one feedback I would offer is that this function is almost identical to the very poorly-named function `linkPlayerResponse` in responses.go further up. It might be a little more work to refactor and make both of them flexible, but in favor of DRY principles, it might be good to find a way to mesh the two and get 1 single function for the functionality

Reply:
I 100% agree with you on that one! I actually tried to overload the function (by having the same name but having different parameters to pass in) but it seems Go doesn't have that as an option. Do you have a recommendation of how I would handle both the information as userID/color pair and as well as guildID/args[0] pair?

Reply:
There's an expression of "map, filter, reduce" in functional programming that applies kinda loosely here. The idea would be to make a general function that accepts the minimum amount of information we need, and in this case, that would probably be userID/color pair. Then for any other use cases that are close, but not quite in that form, we would apply "map, filter, reduce" on top, *to get it into the more general form*. 

So in this example, defining a single function that does what we want with userID/color (the most general type of info we would expect to handle), and then applying whatever filters/transformations needed before calling that function would probably be the cleanest approach. 

So just making a small wrapper function whose sole job is to transform userID/args into userID/color (probably fairly trivial, just some additional parsing to convert the strings to a color) would be a good approach. Let me know if this doesn't make sense, or if you'd like more clarification :)

Reply:
Okay, I think I get what you're saying. However, my first idea of how to implement this after reading through you first 2 paragraphs was to just create one function and have the lines of code that call it filter out the information needed.

But with your last paragraph, I'm guessing your thinking more something along the lines of this:
```
func (guild *GuildState) linkPlayerResponse(s *discordgo.Session, GuildID string, args []string) {

	g, err := s.State.Guild(guild.PersistentGuildData.GuildID)
	if err != nil {
		log.Println(err)
	}

	userID := getMemberFromString(s, GuildID, args[0])
	if userID == "" {
		log.Printf("Sorry, I don't know who `%s` is. You can pass in ID, username, username#XXXX, nickname or @mention", args[0])
	}

	combinedArgs := strings.ToLower(strings.Join(args[1:], ""))

	guild.linkUserByIdAndColor(s, userID, combinedArgs)
}

func (guild *GuildState) linkUserByIdAndColor(s *discordgo.Session, userID string, input string) {

	g, err := s.State.Guild(guild.PersistentGuildData.GuildID)
	if err != nil {
		log.Println(err)
	}

	_, added := guild.checkCacheAndAddUser(g, s, userID)
	if !added {
		log.Println("No users found in Discord for userID " + userID)
	}

	if game.IsColorString(input) {
		// input is a valid color, use it accordingly
		playerData := guild.AmongUsData.GetByColor(input)
		if playerData != nil {
			found := guild.UserData.UpdatePlayerData(userID, playerData)
			if found {
				log.Printf("Successfully linked %s to a color\n", userID)
			} else {
				log.Printf("No player was found with id %s\n", userID)
			}
		}
		return
	} else {
		// input is not a color. Check to see if its a name
		playerData := guild.AmongUsData.GetByName(input)
		if playerData != nil {
			found := guild.UserData.UpdatePlayerData(userID, playerData)
			if found {
				log.Printf("Successfully linked %s by name\n", userID)
			} else {
				log.Printf("No player was found with id %s\n", userID)
			}
		}
	}
}
```

Reply:
I'm thinking this is likely resolve. I'm resolving the conversation for now